### PR TITLE
[Serve] [Docs] Document Serve benchmarks

### LINKED ISSUE #27471: [serve][docs] Benchmark page

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
Signed-off-by: Shreyas Krishnaswamy <shrekris@anyscale.com>

## Why are these changes needed?

Ray Serve's documentation lists some general benchmarks. This change adds benchmarks based on the nightly Serve release tests.

The specific benchmark numbers come from [the nightly run on August 10th, 2022](https://buildkite.com/ray-project/release-tests-branch/builds/885).

Link to the change's "Performance Tuning" page: https://ray--27711.org.readthedocs.build/en/27711/serve/performance.html

## Related issue number

Closes #27471.

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - This change uses the nightly release tests to benchmark Ray Serve.

### CODE REVIEW

File: doc/source/serve/performance.md [hunk 1]
```diff
@@ -20,9 +20,13 @@ We are continuously benchmarking Ray Serve. The metrics we care about are latenc
...
 - It is horizontally scalable so you can add more machines to increase the overall throughput. Ray Serve is built on top of Ray,
-  so its scalability is bounded by Ray’s scalability. Please check out Ray’s [scalability envelope](https://github.com/ray-project/ray/blob/master/release/benchmarks/README.md)
+  so its scalability is bounded by Ray’s scalability. Please see Ray’s [scalability envelope](https://github.com/ray-project/ray/blob/master/release/benchmarks/README.md)
   to learn more about the maximum number of nodes and other limitations.
 
+We run long-running benchmarks nightly. You can check out [our workloads in the Ray repository](https://github.com/ray-project/ray/tree/master/release/serve_tests/workloads).
+
+You can see our nightly benchmark runs [here](https://buildkite.com/ray-project/release-tests-branch/builds?branch=master). They're labeled as "Nightly tests." For example, [here's the nightly benchmark run from July 22nd, 2022](https://buildkite.com/ray-project/release-tests-branch/builds/806). You can see the results by clicking "Serve tests" and choosing a workload.
```
Comment (edoakes) [inline doc/source/serve/performance.md:28]: we can't link to buildkite because it requires people to sign in

File: doc/source/serve/performance.md [hunk 2]
```diff
@@ -20,20 +19,24 @@ We are continuously benchmarking Ray Serve. The metrics we care about are latenc
...
 ## Debugging performance issues
 
 The performance issue you're most likely to encounter is high latency and/or low throughput for requests.
 
-If you have set up [monitoring](serve-monitoring) with Ray and Ray Serve, you will likely observe the following:
+Once you set up [monitoring](serve-monitoring) with Ray and Ray Serve, you can likely observe the following:
 
 - `serve_num_router_requests` is constant while your load increases
-- `serve_deployment_queuing_latency_ms` is spiking up as queries queue up in the background
+- `serve_deployment_processing_latency_ms` is spiking up as queries queue up in the background
```
Comment (shrekris-anyscale) [inline doc/source/serve/performance.md:39]: #15065 removed `serve_deployment_queuing_latency_ms`, and the closest metric I found was `serve_deployment_processing_latency_ms`. @simon-mo @edoakes Does this replacement make sense?

Comment (simon-mo) [reply doc/source/serve/performance.md:39]: Yes. Make sense. Same metrics basically.

Comment (simon-mo) [reply doc/source/serve/performance.md:39]: Yes. I think the metrics is basically queuing + processing.

File: doc/source/serve/performance.md [hunk 3]
```diff
@@ -20,20 +19,132 @@ We are continuously benchmarking Ray Serve. The metrics we care about are latenc
...
+
+```{eval-rst}
+.. list-table::
+   :header-rows: 1
+
+   * - Benchmark
+     - Description
+     - Cluster Details
+     - Performance Numbers
+   * - `Single Deployment <https://github.com/ray-project/ray/blob/master/release/serve_tests/workloads/single_deployment_1k_noop_replica.py>`_
```
Comment (shrekris-anyscale) [inline doc/source/serve/performance.md:35]: Should we specify a commit hash in these links, so these workloads can be updated without creating misleading or dead links?

File: doc/source/serve/performance.md [hunk 4]
```diff
@@ -20,20 +19,132 @@ We are continuously benchmarking Ray Serve. The metrics we care about are latenc
...
+       * per_node_avg_tps = 1.74
+       * per_node_avg_transfer_per_sec_KB = 0.31
+       * cluster_total_thoughput = 1047
+       * cluster_total_transfer_KB = 180.42
+       * cluster_total_timeout_requests = 469.0
+       * cluster_max_P50_latency_ms = 128.97
+       * cluster_max_P75_latency_ms = 239.07
+       * cluster_max_P90_latency_ms = 637.39
+       * cluster_max_P99_latency_ms = 1930.0
+   * - `Serve Handle: Linear Pipeline <https://github.com/ray-project/ray/blob/master/release/serve_tests/workloads/serve_handle_long_chain.py>`_
```
Comment (shrekris-anyscale) [inline doc/source/serve/performance.md:99]: Should we remove the `ServeHandle` and deployment graph benchmarks? They seem to run relatively small workloads.

File: doc/source/serve/performance.md [hunk 5]
```diff
@@ -20,20 +19,132 @@ We are continuously benchmarking Ray Serve. The metrics we care about are latenc
...
+       * throughput_std_tps = 0.43
+       * latency_mean_ms = 126.15
+       * latency_std_ms = 18.35
+```
+
+:::{note}
+The performance numbers above come from a recent run of the nightly benchmarks.
+:::
+
+<!--- See https://github.com/ray-project/ray/pull/27711 for more context on the benchmarks. -->
```
Comment (shrekris-anyscale) [inline doc/source/serve/performance.md:133]: I kept this comment here, so we could track which run these numbers came from via the PR description. Would it be better to include the link directly in this comment?

**Review by shrekris-anyscale — COMMENTED** [review]
I've finished a first draft of the benchmarks. I had a few questions that I put in this review.

**Review by simon-mo — APPROVED** [review]
let's just keep the first two benchmark and select one from deployment graph because
- autoscaling variant gives the same result and doesn't really tell you about performance
- deployment graph result is hard to interrupt without understanding of the topology.

**Comment by shrekris-anyscale** [comment]
> let's just keep the first two benchmark and select one from deployment graph because

That sounds good– I kept [`Deployment Graph: Ensemble`](https://github.com/ray-project/ray/blob/master/release/serve_tests/workloads/deployment_graph_wide_ensemble.py) based on our offline conversation.

File: doc/source/serve/performance.md [hunk 6]
```diff
@@ -20,20 +19,79 @@ We are continuously benchmarking Ray Serve. The metrics we care about are latenc
 - Ray Serve’s latency overhead is single digit milliseconds, around 1-2 milliseconds on average.
 - For throughput, Serve achieves about 3-4k queries per second on a single machine (8 cores) using 1 HTTP proxy actor and 8 replicas performing noop requests.
 - It is horizontally scalable so you can add more machines to increase the overall throughput. Ray Serve is built on top of Ray,
```
Comment (dmatrix) [inline doc/source/serve/performance.md:21]: may be NOP or NOOP or NO-OP requests?

File: doc/source/serve/performance.md (hunk 3 above)
Comment (dmatrix) [reply doc/source/serve/performance.md:35]: whatever we can do to prevent from dead links or 404; slightly out of date more acceptable than 404.

File: doc/source/serve/performance.md [hunk 7]
```diff
@@ -20,20 +19,79 @@ We are continuously benchmarking Ray Serve. The metrics we care about are latenc
...
 
-If you have set up [monitoring](serve-monitoring) with Ray and Ray Serve, you will likely observe the following:
+Once you set up [monitoring](serve-monitoring) with Ray and Ray Serve, these issues may appear as:
 
-- `serve_num_router_requests` is constant while your load increases
-- `serve_deployment_queuing_latency_ms` is spiking up as queries queue up in the background
+- `serve_num_router_requests` staying constant while your load increases
+- `serve_deployment_processing_latency_ms` spiking up as queries queue up in the background
 
 Given these symptoms, there are several ways to fix the issue.
```
Comment (dmatrix) [inline doc/source/serve/performance.md:96]: You suggesting several ways to fix the issue.  But I read only three, and even those are not numbered. Maybe rephrase this as:
"Given these symptoms, there a handful ways to address these issues:

1. Make sure you are using the right hardware and resources
2. Are you providing the GPUs using the (`ray_actor_options={“num_gpus”: 1}`)?
3. Are you using one or more cores (`ray_actor_options={“num_cpus”: 2}`) and setting [`OMP_NUM_THREADS`](serve-omp-num-threads) to increase the performance of your deep learning framework?
4. Consider using `async` methods in your callable. See the section below.
5. Consider batching your requests. See the section below.

Comment (dmatrix) [reply doc/source/serve/performance.md:96]: Sometimes solutions or suggestions offered as list of possibilities is easy on the eye.

File: doc/source/serve/performance.md [hunk 8]
```diff
@@ -42,29 +100,34 @@ Given these symptoms, there are several ways to fix the issue.
...
+then batching is your best approach. Check out the [batching guide](serve-batching) and
+refactor your deployment to accept batches (especially for GPU-based ML inference). You might want to tune `max_batch_size` and `batch_wait_timeout` in the `@serve.batch` decorator to maximize the benefits:
 
 - `max_batch_size` specifies how big the batch should be. Generally,
   we recommend choosing the largest batch size your function can handle
-  AND the performance improvement is no longer sublinear. Take a dummy
-  example: suppose it takes 1ms to process 1 query, 5ms to process 10 queries,
+  without losing the sublinear performance improvement.
+  For example, suppose it takes 1ms to process 1 query, 5ms to process 10 queries,
   and 6ms to process 11 queries. Here you should set the batch size to to 10
```
Comment (dmatrix) [inline doc/source/serve/performance.md:122]: batch size to 10

**Review by dmatrix — APPROVED** [review]
made some nit edits; after which it should be good to go.

File: doc/source/serve/performance.md (hunk 6 above)
Comment (shrekris-anyscale) [reply doc/source/serve/performance.md:21]: Good catch– I replaced it with `no-op`.

File: doc/source/serve/performance.md (hunk 3 above)
Comment (shrekris-anyscale) [reply doc/source/serve/performance.md:35]: Sounds good, I pinned the links using commit hashes. That should help prevent dead links or 404s.

File: doc/source/serve/performance.md (hunk 8 above)
Comment (shrekris-anyscale) [reply doc/source/serve/performance.md:122]: Nice catch, I removed the redundant `to`.

File: doc/source/serve/performance.md (hunk 7 above)
Comment (shrekris-anyscale) [reply doc/source/serve/performance.md]: Good idea– I incorporate a version of that list into the "Debugging performance issues" section. It should be a bit easier to read now.
